### PR TITLE
Improve subtyping check for not yet eta-expanded higher kinded types

### PR DESCRIPTION
**Pull request by prolativ**
Fixes https://github.com/lampepfl/dotty/issues/16183

### CODE REVIEW

**Comment by prolativ** [comment]
The CI is failing because of some infrastructural problems https://github.com/lampepfl/dotty/issues/17119#issuecomment-1479109795 but the compilation tests passed

File: compiler/src/dotty/tools/dotc/core/TypeComparer.scala [hunk 1]
```diff
@@ -580,7 +580,7 @@ class TypeComparer(@constructorOnly initctx: Context) extends ConstraintHandling
             val base = nonExprBaseType(tp1, cls2)
             if (base.typeSymbol == cls2) return true
           }
-          else if tp1.isLambdaSub && !tp1.isAnyKind then
+          else if tp1.typeParams.nonEmpty && !tp1.isAnyKind then
```
[the remark this answers is not in the document]
Comment (prolativ) [reply compiler/src/dotty/tools/dotc/core/TypeComparer.scala:583]: @odersky pushed.
The problem I'm trying to fix here is that when the compiler tries to compare two higher-kinded types like `Foo1` and `Foo2` in the test example, it returns `false` too early without trying to eta-expand the types. In the other cases from the linked issue, where everything compiles successfully, for some reason at least one of the compared types gets already eta-expanded at some earlier stage, which leads the compiler in the right direction to prove that the subtyping relation does occur.
To be more precise, the 2 alternative working workarounds are:
* remove the theoretically redundant package prefix, changing
```scala
implicit val bar: Bar[pkg.Foo2]
```
into
```scala
implicit val bar: Bar[Foo2]
```
* compile both files in a single compiler run (e.g. by renaming `Test_2.scala` to `Test_1.scala` when compiling with vulpix)

[the remark this answers is not in the document]
Comment (prolativ) [reply compiler/src/dotty/tools/dotc/core/TypeComparer.scala:583]: @odersky have you managed to experiment with these changes? If there any chance to push this forward?

**Review by odersky — APPROVED** [review]
After more review, I agree this is the right change.

File: compiler/src/dotty/tools/dotc/core/TypeComparer.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (odersky) [reply compiler/src/dotty/tools/dotc/core/TypeComparer.scala:583]: This does avoid the fourthTry logic for the cases where tp1 is a generic class without type parameters. But I checked that this logic does not do anything in this case, so I think the change is OK.

**Pull request merged**